### ISSUE TITLE
How to enable VAAPI? (ffmpeg --enable-vaapi ?)

### ISSUE THREAD

**Issue by Rom888** [UltraStar-Deluxe/USDX#556]
I tried to enable VAAPI (hardware-accelerated codec) on my ubuntu 20.04 focal because MPV can use it on my PC.
This was added to config.ini as @cRaZy-bisCuiT suggested:
```
[Graphics]
PreferredCodecs=vaapi
```

I think I need to build ffmpeg with --enable-vaapi option.
Where should I put it before building USDX?

At the moment I have this error in the Error.log:
```
ERROR:  Unknown preferred codec vaapi [TVideo_FFmpeg]
INFO:   Valid non-default codecs are: [TVideo_FFmpeg]
INFO:       h263_v4l2m2m [H.263 / H.263-1996, H.263+ / H.263-1998 / H.263 version 2]
INFO:       h264_v4l2m2m [H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10]
INFO:       hevc_v4l2m2m [H.265 / HEVC (High Efficiency Video Coding)]
INFO:       mpeg4_v4l2m2m [MPEG-4 part 2]
INFO:       mpegvideo [MPEG-2 video]
INFO:       mpeg1_v4l2m2m [MPEG-1 video]
INFO:       mpeg2_v4l2m2m [MPEG-2 video]
INFO:       vc1_v4l2m2m [SMPTE VC-1]
INFO:       vp8_v4l2m2m [On2 VP8]
INFO:       vp9_v4l2m2m [Google VP9]
INFO:       libopenjpeg [JPEG 2000]
INFO:       libvpx [On2 VP8]
INFO:       libvpx-vp9 [Google VP9]
INFO:       h264_cuvid [H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10]
INFO:       hevc_cuvid [H.265 / HEVC (High Efficiency Video Coding)]
INFO:       mjpeg_cuvid [Motion JPEG]
INFO:       mpeg1_cuvid [MPEG-1 video]
INFO:       mpeg2_cuvid [MPEG-2 video]
INFO:       mpeg4_cuvid [MPEG-4 part 2]
INFO:       vc1_cuvid [SMPTE VC-1]
INFO:       vp8_cuvid [On2 VP8]
INFO:       vp9_cuvid [Google VP9]
INFO:   VideoStreamIndex : -1 [TVideoPlayback_ffmpeg.Open]
INFO:   Using h264 codec [TVideoPlayback_ffmpeg.Open]
INFO:   Framerate: 24fps [TVideoPlayback_ffmpeg.Open]
INFO:   Using decoder FFmpeg_Decoder for "/home..." [TAudioPlaybackBase.OpenDecodeStream]
```
@andy5995

**Comment by s09bQ5** [comment]
Sorry, VAAPI is not supported.
Supporting the Raspberry Pi decoder hardware was easy because it behaves like a normal software codec that renders an RGB image into RAM. But with VAAPI we have to use a different API in FFmpeg to get an RGB image and the most efficient way to render the image would be to skip that conversion step and instead use a special libva function to let the GPU copy the image into an OpenGL texture.

**Comment by Rom888** [comment]
@s09bQ5 thanks for reply,
Do you think there is any other way to switch on hardware decoding for Intel Graphics Controllers with USDX on Ubuntu 20.04 focal?
My particular case:
```
$ lspci | grep VGA
00:02.0 VGA compatible controller: Intel Corporation 2nd Generation Core Processor Family Integrated Graphics Controller (rev 09)
```

**Comment by s09bQ5** [comment]
The only way is to sit down, make oneself familiar with the FFmpeg API for hardware acceleration, and change the source code of USDX to use it.

**Comment by dgruss** [comment]
this is not supported/planned --> should be closed

**Comment by barbeque-squared** [comment]
In principle agree, though I do have questions about what this `PreferredCodecs` thing is supposed to be for anyway? This feels like yet another hidden setting I'm only now learning about, which still appears to have stuff lying around in the source code.

Obviously not related to this issue itself, but keep it open for a day or two until I can create a follow-up issue to pose my questions (the gist of my question will be "what is this even for and do we need to keep it around? if yes, it should be documented, if no, it should be nuked")

**Comment by dgruss** [comment]
I'd still like to make progress on this issue and I take the comment on #1329 as a decision to implement this feature?

My understanding is that this will be a Linux-exclusive feature with about 5000-10000 lines of code? We probably want zero-copy rendering; copying every decoded frame back to RAM and then uploading it again likely defeats the entire point of supporting VAAPI, correct?

We should silently fall back to software decoding if VAAPI is unavailable or fails. A/V sync may turn out to become a problem but we will see...

Sounds like a small change, maybe 5k-10k LOC, I will start looking into it.

**Comment by dgruss** [comment]
I looked into this a bit and it appears to be less code necessary... about 1000 loc to get vaapi support with CPU transfer mode. savings on a 10 year old laptop playing a 4k h264 video, about 20% (from 300% to 240% cpu time while playing).
With zero copy, we add 500 loc for this specific codec + hw combination, which brings down the CPU usage drastically to around 10-20% cpu time while playing, which is awesome, but: for instance rpi4/5 requires further code to support the codecs + hw combination that rpi4/5 needs. So for every such combination, we might have to add a few hundred lines, maybe another 300-500 or so.

A different approach could be to switch video playback entirely to libmpv. That would allow keeping all these hardware aspects outside of the game as libmpv takes care of it. There we are at about 1500 loc with full support on any hardware/codec combination that libmpv supports.

**Comment by dgruss** [comment]
I just realized that @Rom888 already had mentioned mpv - this might be the right path forward here instead of supporting vaapi directly in the game.

**Comment by s09bQ5** [comment]
> but: for instance rpi4/5 requires further code to support the codecs + hw combination that rpi4/5 needs

Their decoders are V4L2 devices that are not tied to the GPU. They should already be usable with PreferredCodecs using h264_v4l2m2m or hevc_v4l2m2m. The only drawback is that this is not zero copy.

**Comment by dgruss** [comment]
Exactly. But the real gain on that Intel laptop i only saw with zero copy. Libmpv would be zero copy though...
